### PR TITLE
Authorize empty content put in custom operations

### PR DESCRIPTION
**Pull request by thomasglachant**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | no
| License       | MIT
| Doc PR        | NA

This PR allows to send empty request content for PUT in a custom operation.

Use case : 
I have a custom Action (ex : "reset") associated to an ApiResource "Job". 
I don't have any request content to send when I send a PUT request to `/jobs/1234/reset`.
But I've a SerializerError because the request content is required.

### CODE REVIEW

File: src/EventListener/DeserializeListener.php [hunk 1]
```diff
@@ -51,6 +51,7 @@ public function onKernelRequest(GetResponseEvent $event)
             || $request->isMethod(Request::METHOD_DELETE)
             || !($attributes = RequestAttributesExtractor::extractAttributes($request))
             || !$attributes['receive']
+            || ($request->isMethod(Request::METHOD_PUT) && $request->getContent() === '')
```
Comment (dunglas) [inline src/EventListener/DeserializeListener.php]: Can you store the result of `$request->getContent()` in a variable to avoid 2 (costly) method calls (the next one is line 70)?
Also, please use Yoda styl (`'' === $content`).

File: src/EventListener/DeserializeListener.php [hunk 2]
```diff
@@ -52,7 +52,7 @@ public function onKernelRequest(GetResponseEvent $event)
             || $request->isMethod(Request::METHOD_DELETE)
             || !($attributes = RequestAttributesExtractor::extractAttributes($request))
             || !$attributes['receive']
-            || ($request->isMethod(Request::METHOD_PUT) &&  '' === $requestContent)
+            || ($request->isMethod(Request::METHOD_PUT) && '' === $requestContent)
```
Comment (dunglas) [inline src/EventListener/DeserializeListener.php]: May I suggest:

`|| ($request->isMethod(Request::METHOD_PUT) && '' === ($requestContent = $request->getContent()))`

It will avoid to always retrieve the body (that can be huge) in most cases.

**Comment by meyerbaptiste** [comment]
> in custom operation

But in reality it will work for all `PUT` operations, custom or not. Do we really want this?

**Comment by dunglas** [comment]
What's the behavior with this patch when an empty request is sent to a standard PUT route?

File: src/EventListener/DeserializeListener.php (hunk 2 above)
Comment (thomasglachant) [inline src/EventListener/DeserializeListener.php]: Ok, I can do it, but it's a little more complicated because it's a "&&" and for a POST we don't initialize the variable. 
So I need to add `$requestContent ?? $request->getContent()` in the deserialize line.

It will be faster but less clear.

**Comment by meyerbaptiste** [comment]
We have a 200 with the unmodified resource instead of a 400.

**Comment by thomasglachant** [comment]
> What's the behavior with this patch when an empty request is sent to a standard PUT route?

Nothing (200) ... I wrote a behat test to check. It's like a PUT request with an empty array in body.

File: src/EventListener/DeserializeListener.php (hunk 2 above)
Comment (dunglas) [inline src/EventListener/DeserializeListener.php]: As this listener will be executed for every request (even if its a page not managed by API Platform), let's use the fastest alternative.

**Comment by dunglas** [comment]
It looks acceptable to me. WDYT @api-platform/core-team?

**Comment by soyuka** [comment]
Ok to me.

File: src/EventListener/DeserializeListener.php (hunk 2 above)
Comment (thomasglachant) [inline src/EventListener/DeserializeListener.php]: done ! I add a behat test with the empty body ... don't know if it's necessary ...

**Comment by meyerbaptiste** [comment]
It looks like a BC break to have a 200 instead of a 400 now. Nop?

**Comment by soyuka** [comment]
Lot of overhead if we need to check that it's a custom operation no?

**Comment by dunglas** [comment]
It's borderline, but it looks acceptable. Who would rely on an empty request throwing a 400?

File: src/EventListener/DeserializeListener.php [hunk 3]
```diff
@@ -46,11 +46,13 @@ public function __construct(SerializerInterface $serializer, SerializerContextBu
     public function onKernelRequest(GetResponseEvent $event)
     {
         $request = $event->getRequest();
+        $requestContent = null;
         if (
             $request->isMethodSafe(false)
             || $request->isMethod(Request::METHOD_DELETE)
             || !($attributes = RequestAttributesExtractor::extractAttributes($request))
             || !$attributes['receive']
+            || ($request->isMethod(Request::METHOD_PUT) && '' === ($requestContent = $request->getContent()))
```
Comment (dunglas) [inline src/EventListener/DeserializeListener.php]: If you invert the conditions (`'' === ($requestContent = $request->getContent())` && $request->isMethod(Request::METHOD_PUT)`), you don't need the trick with `??` anymore because the variable will always be defined. WDYT?

Comment (thomasglachant) [inline src/EventListener/DeserializeListener.php]: Good idea, clean and fast 👍

File: features/main/crud.feature [hunk 4]
```diff
@@ -395,6 +395,41 @@ Feature: Create-Retrieve-Update-Delete
     }
     """
 
+  Scenario: Update a resource with empty body
```
Comment (thomasglachant) [inline features/main/crud.feature]: @dunglas Do you think it's necessary to add this test ? I'm not convinced ...

Comment (soyuka) [inline features/main/crud.feature]: There are no useless integration test ;).

**Comment by Simperfit** [comment]
It's not like someone should rely on a 400 return. LGTM

**Comment by soyuka** [comment]
Thanks @thomasglachant !

**Comment by teohhanhui** [comment]
I really do not understand why this needs to be specially handled. What happened to the PR(s) on allowing empty request body?

**Comment by soyuka** [comment]
Empty request body does make sense only with PUT / DELETE requests no?

**Comment by teohhanhui** [comment]
No. See #782 and #805.
